### PR TITLE
[UNR-3547][MS] RPCs are acked when they have actually been processed. 

### PR DESCRIPTION
**Pull request by MatthewSandfordImprobable**
#### Description
RPCs will now only be acknowledged if the they have been processed and removed from the RPC queue.
Additionally before applying an RPC we now check the role of the actor. If we do not have authority then we do not apply the RPC but will continue to process queued RPCs.

Testing: Testing on the offloading benchmark gym and the example project capture point map.

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialRPCService.cpp [hunk 1]
```diff
@@ -427,7 +427,7 @@ void SpatialRPCService::ExtractRPCsForType(Worker_EntityId EntityId, ERPCType Ty
 			const TOptional<RPCPayload>& Element = Buffer.GetRingBufferElement(RPCId);
 			if (Element.IsSet())
 			{
-				bool bKeepExtracting = ExtractRPCCallback.Execute(EntityId, Type, Element.GetValue());
+				bool bKeepExtracting = ExtractRPCCallback.Execute(EntityId, Type, Element.GetValue(), RPCId);
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialRPCService.cpp:430]:
```suggestion
				const bool bKeepExtracting = ExtractRPCCallback.Execute(EntityId, Type, Element.GetValue(), RPCId);
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 2]
```diff
@@ -1950,14 +1950,29 @@ ERPCResult USpatialReceiver::ApplyRPCInternal(UObject* TargetObject, UFunction*
...
+	if (UnresolvedRefs.Num() != 0 && !bApplyWithUnresolvedRefs)
 	{
-		TargetObject->ProcessEvent(Function, Parms);
-		Result = ERPCResult::Success;
+		ErrorInfo.ErrorCode = ERPCResult::UnresolvedParameters;
 	}
 	else
 	{
-		Result = ERPCResult::UnresolvedParameters;
+		// Get the supposedly net owned actor
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:1959]: NIT, might be misunderstanding this comment, but personally think this wasn't explicit enough before?

```suggestion
		// Get the RPC target Actor.
```

File: SpatialGDK/Source/SpatialGDK/Public/SpatialConstants.h [hunk 3]
```diff
@@ -292,6 +292,8 @@ const FString SCHEMA_DATABASE_ASSET_PATH = TEXT("/Game/Spatial/SchemaDatabase");
 
 const FString DEV_LOGIN_TAG = TEXT("dev_login");
 
+const int64 INVALID_RPC_ID = -1;
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Public/SpatialConstants.h:295]:
```suggestion
const uint64 INVALID_RPC_ID = -1;
```

File: SpatialGDK/Source/SpatialGDK/Private/Utils/RPCContainer.cpp [hunk 4]
```diff
@@ -193,11 +205,17 @@ bool FRPCContainer::ApplyFunction(FPendingRPCParams& Params)
...
-		LogRPCError(ErrorInfo, QueueType, Params);
+	LogRPCError(ErrorInfo, QueueType, Params);
 #endif
-		return ErrorInfo.bShouldDrop;
+	return ErrorInfo.bShouldDrop;
+}
+
+void FRPCContainer::ExecuteRPCQueueProcessingUpdateFunction(const FPendingRPCParams& Params)
+{
+	if (RPCQueueProcessingUpdateDelegate.IsBound())
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/RPCContainer.cpp:217]: Can we check this? Seems like a bit of a death sentence for the GDK to not have this bound

File: SpatialGDK/Source/SpatialGDK/Public/SpatialConstants.h (hunk 3 above)
Comment (MatthewSandfordImprobable) [reply SpatialGDK/Source/SpatialGDK/Public/SpatialConstants.h:295]: I think we should defo make it uint64, that was an error on my part. 
I think we should use -1 as I believe this will set INVALID_RPC_ID to the highest possible value of an uint64 which is probably safer than 0.

Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Public/SpatialConstants.h:295]: 0 is an invalid rpc ID though. uint64_max is valid.

Comment (MatthewSandfordImprobable) [reply SpatialGDK/Source/SpatialGDK/Public/SpatialConstants.h:295]: True, this a minor detail and I do not care either way but... I think it is more likely that 0 becomes a valid rpc ID at some point than we ever reach uint64_max rpcs. If we do reach uint64_max we have a bigger problem no?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 5]
```diff
@@ -1950,14 +1950,29 @@ ERPCResult USpatialReceiver::ApplyRPCInternal(UObject* TargetObject, UFunction*
...
 	}
 	else
 	{
-		Result = ERPCResult::UnresolvedParameters;
+		// Get the RPC target Actor.
+		AActor* Actor = Cast<AActor>(TargetObject);
+		if (Actor == nullptr)
+		{
+			Actor = Cast<UActorComponent>(TargetObject)->GetOwner();
+		}
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:1964]: This will not work for subobjects that aren't `UActorComponent`s. I believe this is what we use in other places:
```suggestion
		AActor* Actor = TargetObject->GetTypedOuter<AActor>();
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 6]
```diff
@@ -1935,29 +1934,44 @@ void USpatialReceiver::ApplyComponentUpdate(const Worker_ComponentUpdate& Compon
...
 	else
 	{
-		Result = ERPCResult::UnresolvedParameters;
+		// Get the RPC target Actor.
+		AActor* Actor = TargetObject->GetTypedOuter<AActor>();
+
+		const bool bHasServerEndpointAuthority = StaticComponentView->HasAuthority(Params.ObjectRef.Entity, SpatialConstants::SERVER_ENDPOINT_COMPONENT_ID);
+		if (bHasServerEndpointAuthority &&
+			Actor->Role == ROLE_SimulatedProxy &&
+			Params.Type != ERPCType::NetMulticast)
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:1964]: I believe we need this condition instead. We want to catch cases where the references got resolved (or timed out) when the server knows it should no longer have authority (so it set the role to `SimulatedProxy`), but it may or may not still have authority on the server endpoint.
```suggestion
		if ((Params.Type == ERPCType::ServerReliable ||
			Params.Type == ERPCType::ServerUnreliable) &&
			Actor->Role == ROLE_SimulatedProxy)
```

File: SpatialGDK/Source/SpatialGDK/Private/Utils/RPCContainer.cpp [hunk 7]
```diff
@@ -114,15 +117,19 @@ void FRPCContainer::ProcessRPCs(FArrayOfParams& RPCList)
...
 		{
+		case ERPCQueueCommand::ContinueProcessing:
 			NumProcessedParams++;
-		}
-		else
-		{
+		case ERPCQueueCommand::DropEntireQueue:
+			RPCList.Empty();
+			return;
+		case ERPCQueueCommand::DropRPC:
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/RPCContainer.cpp:128]: This doesn't look right. Previously, when `bShouldDrop` was true, `ApplyFunction` would return `true` so this would keep processing the RPCs. We really have these cases:
* RPC executed successfully, keep processing
* RPC failed, keep processing (drop RPC)
* RPC failed, queue it
* RPC failed, drop entire queue

Comment (improbable-valentyn) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/RPCContainer.cpp:128]: In fact, for dropping RPC on the sending side, it looks like we would actually end up dropping the entire queue in most cases:
* target object was destroyed
* we don't have an actor channel
* (for legacy RPCs) we don't have authority and `WillHaveAuthorityOverActor` returns false
The only remaining case is `MissingFunctionInfo`, and that one looks quite odd to me. I can't imagine a situation where you call an RPC but the function is missing. I guess we can just drop it and keep processing the remaining RPCs in the queue just in case.

**Comment by MatthewSandfordImprobable** [comment]
Updated the regression on the sending side. I believe this fixes the issue.

File: SpatialGDK/Source/SpatialGDK/Public/SpatialConstants.h (hunk 3 above)
Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Public/SpatialConstants.h]: I think it's impossible the 0 can ever be a valid RPC ID unless we make the ID an option field, which seems unlikely.

But trying to make this -1 means if you actually compare invalid ID to the actually invalid ID that can be stored in schema it will appear different.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp [hunk 8]
```diff
@@ -736,13 +736,15 @@ FRPCErrorInfo USpatialSender::SendLegacyRPC(UObject* TargetObject, UFunction* Fu
...
 		{
-			bShouldDrop = !WillHaveAuthorityOverActor(TargetActor, TargetObjectRef.Entity);
+			if (!WillHaveAuthorityOverActor(TargetActor, TargetObjectRef.Entity))
+			{
+				QueueCommand = ERPCQueueCommand::StopProcessing;
+			}
 		}
-
-		return FRPCErrorInfo{ TargetObject, Function, ERPCResult::NoAuthority, bShouldDrop };
+		return FRPCErrorInfo{ TargetObject, Function, ERPCResult::NoAuthority, QueueCommand };
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp:747]: Previously, if `WillHaveAuthorityOverActor` was false, it would drop the RPC and keep processing the queue, now it seems to stop processing and instead queues this RPC, which is incorrect. I believe in this case we should also drop the entire queue, as it should be impossible for this worker to gain authority in between processing these RPCs, so it will drop all of them.

File: SpatialGDK/Source/SpatialGDK/Private/Utils/RPCContainer.cpp [hunk 9]
```diff
@@ -187,20 +194,19 @@ void FRPCContainer::BindProcessingFunction(const FProcessRPCDelegate& Function)
...
-	{
+
 #if !UE_BUILD_SHIPPING
-		LogRPCError(ErrorInfo, QueueType, Params);
+	LogRPCError(ErrorInfo, QueueType, Params);
 #endif
-		return ErrorInfo.bShouldDrop;
-	}
+
+	return ErrorInfo.QueueCommand;
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/RPCContainer.cpp:211]: We only seem to set this to `StopProcessing` in `SendLegacyRPC` (which I left a comment about), but previously, any RPC that had a result other than `Success` would get queued, so perhaps the default value for `QueueCommand` should be `StopProcessing`?

Comment (improbable-valentyn) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/RPCContainer.cpp:211]: In fact, `QueueCommand` is only used when the result is not success, so perhaps we could reflect that in the name? e.g. `QueueCommandOnFailure`

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp [hunk 10]
```diff
@@ -629,21 +629,21 @@ FRPCErrorInfo USpatialSender::SendRPC(const FPendingRPCParams& Params)
 	if (!TargetObjectWeakPtr.IsValid())
 	{
 		// Target object was destroyed before the RPC could be (re)sent
-		return FRPCErrorInfo{ nullptr, nullptr, ERPCResult::UnresolvedTargetObject, true };
+		return FRPCErrorInfo{ nullptr, nullptr, ERPCResult::UnresolvedTargetObject, ERPCQueueCommand::ContinueProcessing };
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp:632]:
```suggestion
		return FRPCErrorInfo{ nullptr, nullptr, ERPCResult::UnresolvedTargetObject, ERPCQueueCommand::DropEntireQueue };
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp [hunk 11]
```diff
@@ -629,21 +629,21 @@ FRPCErrorInfo USpatialSender::SendRPC(const FPendingRPCParams& Params)
...
 	{
-		return FRPCErrorInfo{ TargetObject, nullptr, ERPCResult::MissingFunctionInfo, true };
+		return FRPCErrorInfo{ TargetObject, nullptr, ERPCResult::MissingFunctionInfo, ERPCQueueCommand::ContinueProcessing };
 	}
 
 	USpatialActorChannel* Channel = NetDriver->GetOrCreateSpatialActorChannel(TargetObject);
 	if (Channel == nullptr)
 	{
-		return FRPCErrorInfo{ TargetObject, Function, ERPCResult::NoActorChannel, true };
+		return FRPCErrorInfo{ TargetObject, Function, ERPCResult::NoActorChannel, ERPCQueueCommand::ContinueProcessing };
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp:646]:
```suggestion
		return FRPCErrorInfo{ TargetObject, Function, ERPCResult::NoActorChannel, ERPCQueueCommand::DropEntireQueue };
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp (hunk 10 above)
Comment (improbable-valentyn) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp:632]: If the object was destroyed, all the other RPCs queued up for it will fail as well so might as well drop them.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp (hunk 11 above)
Comment (improbable-valentyn) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp:646]: If the channel is missing, all the other RPCs queued up for this object will fail as well so might as well drop them.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp [hunk 12]
```diff
@@ -814,20 +816,20 @@ void USpatialSender::TrackRPC(AActor* Actor, UFunction* Function, const RPCPaylo
...
 			{
-				WillHaveAuthorityOverActor = false;
+				bWillHaveAuthorityOverActor = false;
 			}
 		}
 	}
 
-	return WillHaveAuthorityOverActor;
+	return bWillHaveAuthorityOverActor;
 }
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp:833]: I think this whole method can be nuked, it existed for the time when we briefly set Actor role to SimulatedProxy when the entity was being created - this isn't the case anymore. Should be able to replace all occurrences of this function called with `Actor->HasAuthority()` - would you mind drive-by fixing this?

File: SpatialGDK/Source/SpatialGDK/Private/Tests/RPCServiceTest.cpp [hunk 13]
```diff
@@ -1,11 +1,18 @@
...
+#include "Interop/SpatialReceiver.h"
 #include "Interop/SpatialRPCService.h"
 #include "Interop/SpatialStaticComponentView.h"
+#include "Misc/AutomationTest.h"
 #include "Schema/RPCPayload.h"
 #include "SpatialConstants.h"
 #include "SpatialGDKSettings.h"
+#include "Tests/TestActor.h"
+#include "Tests/AutomationCommon.h"
+#include "Tests/AutomationEditorCommon.h"
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/Tests/RPCServiceTest.cpp:15]: nice alphabet bro

File: SpatialGDK/Source/SpatialGDK/Private/Tests/RPCServiceTest.cpp [hunk 14]
```diff
@@ -1,11 +1,18 @@
...
 #include "CoreMinimal.h"
+#include "Engine/Engine.h"
+#include "GameFramework/GameStateBase.h"
+#include "Interop/SpatialReceiver.h"
 #include "Interop/SpatialRPCService.h"
 #include "Interop/SpatialStaticComponentView.h"
+#include "Misc/AutomationTest.h"
 #include "Schema/RPCPayload.h"
 #include "SpatialConstants.h"
 #include "SpatialGDKSettings.h"
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/Tests/RPCServiceTest.cpp:12]: Can you drive-by split these GDK / engine imports into separate groups? (worker on using clang-format and it sorts per group so files like this will get rekt)

File: SpatialGDK/Source/SpatialGDK/Private/Utils/RPCContainer.cpp [hunk 15]
```diff
@@ -101,8 +101,11 @@ void FRPCContainer::ProcessOrQueueRPC(const FUnrealObjectRef& TargetObjectRef, E
 
 	if (!ObjectHasRPCsQueuedOfType(Params.ObjectRef.Entity, Params.Type))
 	{
-		if (ApplyFunction(Params))
+		ERPCQueueCommand QueueCommand = ApplyFunction(Params);
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/RPCContainer.cpp:104]:
```suggestion
		const ERPCQueueCommand QueueCommand = ApplyFunction(Params);
```

File: SpatialGDK/Source/SpatialGDK/Private/Utils/RPCContainer.cpp [hunk 16]
```diff
@@ -117,15 +120,19 @@ void FRPCContainer::ProcessRPCs(FArrayOfParams& RPCList)
 	int NumProcessedParams = 0;
 	for (auto& Params : RPCList)
 	{
-		if (ApplyFunction(Params))
+		ERPCQueueCommand QueueCommand = ApplyFunction(Params);
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/RPCContainer.cpp:123]:
```suggestion
		const ERPCQueueCommand QueueCommand = ApplyFunction(Params);
```

File: SpatialGDK/Source/SpatialGDK/Public/Tests/TestActor.h [hunk 17]
```diff
@@ -0,0 +1,24 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#pragma once
+
+#include "CoreMinimal.h"
+#include "GameFramework/Actor.h"
+#include "TestActor.generated.h"
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Public/Tests/TestActor.h:7]: NIT: can u split the `TestActor.generated.h` to next line?

File: SpatialGDK/Source/SpatialGDK/Public/Tests/TestActor.h [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+{
+	GENERATED_BODY()
+
+	ATestActor()
+	{
+		SetReplicates(true);
+	}
+
+public:
+
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Public/Tests/TestActor.h:20]:
```suggestion
```

File: SpatialGDK/Source/SpatialGDK/Public/Utils/RPCContainer.h [hunk 19]
```diff
@@ -45,6 +45,13 @@ enum class ERPCResult : uint8
 	Unknown
 };
 
+enum class ERPCQueueCommand : uint8_t
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Public/Utils/RPCContainer.h:48]: NIT - made a bit more sense as this when I was reading, what you think?
```suggestion
enum class ERPCQueueResult : uint8_t
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 20]
```diff
@@ -1968,15 +1982,15 @@ FRPCErrorInfo USpatialReceiver::ApplyRPC(const FPendingRPCParams& Params)
 	TWeakObjectPtr<UObject> TargetObjectWeakPtr = PackageMap->GetObjectFromUnrealObjectRef(Params.ObjectRef);
 	if (!TargetObjectWeakPtr.IsValid())
 	{
-		return FRPCErrorInfo{ nullptr, nullptr, ERPCResult::UnresolvedTargetObject };
+		return FRPCErrorInfo{ nullptr, nullptr, ERPCResult::UnresolvedTargetObject, ERPCQueueCommand::StopProcessing };
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:1985]: When does this conditional get hit / why would we want this response? (vs. deleting entire list)

File: SpatialGDK/Source/SpatialGDK/Private/Tests/RPCServiceTest.cpp [hunk 21]
```diff
@@ -496,3 +508,128 @@ RPC_SERVICE_TEST(GIVEN_receiving_an_rpc_WHEN_return_false_from_extract_callback_
...
+	if (!ensure(Test != nullptr))
+	{
+		return true;
+	}
+
+	if (!ensure(Data != nullptr) ||
+		!ensure(Data->Actor != nullptr))
+	{
+		Test->TestTrue("Returning false in extraction callback correctly stopped processing RPCs", false);
+	}
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/Tests/RPCServiceTest.cpp:572]: I'm kinda confused by these / not seen this pattern before. Can we use `check(...)` instead or will that have different behaviour?

File: SpatialGDK/Source/SpatialGDK/Public/Utils/RPCContainer.h (hunk 19 above)
Comment (MatthewSandfordImprobable) [reply SpatialGDK/Source/SpatialGDK/Public/Utils/RPCContainer.h:48]: Hmm, that seems more confusing to me. A ERPCQueueResult to me suggests what is returned after we attempt to queue something. This is used mostly for processing the queue. Maybe ERPCQueueProcessResult.

File: SpatialGDK/Source/SpatialGDK/Private/Tests/RPCServiceTest.cpp (hunk 21 above)
Comment (MatthewSandfordImprobable) [reply SpatialGDK/Source/SpatialGDK/Private/Tests/RPCServiceTest.cpp:572]: Checks are fine but then we would crash after. As you can continue after checks I use ensures which will return whether the check was successful. This allows you to return from code that will crash.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp (hunk 20 above)
Comment (MatthewSandfordImprobable) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:1985]: I am not an expert on this but... talking with Valentyn he suggested maybe there is a case with dynamic sub objects that would hit this condition. Perhaps the thought being we might later receive the sub object and process the RPC on it.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp (hunk 12 above)
Comment (MatthewSandfordImprobable) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp:833]: Let me know if you think the changes I have made make sense. We might be able to remove the new TargetActor->HasAuthority() check as it might be redundant after NetDriver->StaticComponentView->HasAuthority(EntityId, ComponentId).

File: SpatialGDK/Source/SpatialGDK/Private/Tests/RPCServiceTest.cpp [hunk 22]
```diff
@@ -496,3 +508,128 @@ RPC_SERVICE_TEST(GIVEN_receiving_an_rpc_WHEN_return_false_from_extract_callback_
...
+	TestingComponentViewHelpers::AddEntityComponentToStaticComponentView(*SpatialNetDriver->StaticComponentView,
+		EntityId, SpatialConstants::SERVER_ENDPOINT_COMPONENT_ID,
+		GetServerAuthorityFromRPCEndpointType(EndpointType));
+
+	SpatialGDK::SpatialRPCService* RPCService = SpatialNetDriver->GetRPCService();
+	RPCService->OnEndpointAuthorityGained(EntityId, SpatialConstants::SERVER_ENDPOINT_COMPONENT_ID);
+
+	bool bSuccess = false;
+	FProcessRPCDelegate RPCDelegate = FProcessRPCDelegate::CreateLambda([&SpatialNetDriver, &bSuccess](const FPendingRPCParams& Params) {
+
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/Tests/RPCServiceTest.cpp:600]:
```suggestion
	FProcessRPCDelegate RPCDelegate = FProcessRPCDelegate::CreateLambda([&SpatialNetDriver, &bSuccess](const FPendingRPCParams& Params) 
	{
```

File: SpatialGDK/Source/SpatialGDK/Private/Tests/RPCServiceTest.cpp [hunk 23]
```diff
@@ -496,3 +508,128 @@ RPC_SERVICE_TEST(GIVEN_receiving_an_rpc_WHEN_return_false_from_extract_callback_
...
+DEFINE_LATENT_AUTOMATION_COMMAND_ONE_PARAMETER(FWaitForActor, TSharedPtr<TestData>, Data);
+bool FWaitForActor::Update()
+{
+	AActor* Actor = Data->Actor;
+	return (IsValid(Actor) && Actor->IsActorInitialized() && Actor->HasActorBegunPlay());
+}
+
+DEFINE_LATENT_AUTOMATION_COMMAND_TWO_PARAMETER(FDropRPCQueueTest, FAutomationTestBase*, Test, TSharedPtr<TestData>, Data);
+bool FDropRPCQueueTest::Update()
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/Tests/RPCServiceTest.cpp:565]: Can you refactor this down into separate steps rather than the whole test in one?

File: SpatialGDK/Source/SpatialGDK/Private/Tests/RPCServiceTest.cpp [hunk 24]
```diff
@@ -496,3 +508,128 @@ RPC_SERVICE_TEST(GIVEN_receiving_an_rpc_WHEN_return_false_from_extract_callback_
...
+		ClientComponentData,
+		GetClientAuthorityFromRPCEndpointType(EndpointType));
+	TestingComponentViewHelpers::AddEntityComponentToStaticComponentView(*SpatialNetDriver->StaticComponentView,
+		EntityId, SpatialConstants::SERVER_ENDPOINT_COMPONENT_ID,
+		GetServerAuthorityFromRPCEndpointType(EndpointType));
+
+	SpatialGDK::SpatialRPCService* RPCService = SpatialNetDriver->GetRPCService();
+	RPCService->OnEndpointAuthorityGained(EntityId, SpatialConstants::SERVER_ENDPOINT_COMPONENT_ID);
+
+	bool bSuccess = false;
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/Tests/RPCServiceTest.cpp:602]: Can rename to something more descriptive like `bRpcFailedDueToNoAuthority`

File: SpatialGDK/Source/SpatialGDK/Private/Tests/RPCServiceTest.cpp (hunk 23 above)
Comment (MatthewSandfordImprobable) [reply SpatialGDK/Source/SpatialGDK/Private/Tests/RPCServiceTest.cpp:565]: Is that necessary? The reason these are separated is for temporal reasons no?

Comment (alastairdglennie) [reply SpatialGDK/Source/SpatialGDK/Private/Tests/RPCServiceTest.cpp:513]: Yea fair, I guess not in this case, I think if we want to add more tests then more modular steps would help a lot but can delay for now if you pref

File: SpatialGDK/Source/SpatialGDK/Public/Utils/RPCContainer.h (hunk 19 above)
Comment (alastairdglennie) [reply SpatialGDK/Source/SpatialGDK/Public/Utils/RPCContainer.h:48]: Sure, I think both are more descriptive than `ERPCQueueCommand`

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 25]
```diff
@@ -1964,7 +1961,24 @@ ERPCResult USpatialReceiver::ApplyRPCInternal(UObject* TargetObject, UFunction*
...
+		AActor* Actor = TargetObject->IsA<AActor>() ? Cast<AActor>(TargetObject) : TargetObject->GetTypedOuter<AActor>();
+		ERPCType RPCType = PendingRPCParams.Type;
+
+		if (Actor->Role == ROLE_SimulatedProxy &&
+			(RPCType == ERPCType::ServerReliable ||
+			 RPCType == ERPCType::ServerUnreliable))
+		{
+			ErrorInfo.ErrorCode = ERPCResult::NoAuthority;
+			ErrorInfo.QueueProcessResult = ERPCQueueProcessResult::DropEntireQueue;
+			UE_LOG(LogSpatialReceiver, Warning, TEXT("Trying to execute Server RPC %s::%s without authority over the target actor. RPCs for this actor will be dropped."), *GetNameSafe(TargetObject), *GetNameSafe(Function));
```
[the remark this answers is not in the document]
Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:1974]: Do we drop server RPCs invoked on a non auth server (I know on the auth server we just invoke locally)

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialRPCService.cpp [hunk 26]
```diff
@@ -464,13 +473,30 @@ void SpatialRPCService::ExtractRPCsForType(Worker_EntityId EntityId, ERPCType Ty
...
-			const EntityComponentId EntityComponentPair = { EntityId, RPCRingBufferUtils::GetAckComponentId(Type) };
+			LastSeenRPCIds[EntityTypePair] = LastProcessedRPCId;
+		}
+	}
+}
 
-			Schema_Object* EndpointObject = Schema_GetComponentUpdateFields(GetOrCreateComponentUpdate(EntityComponentPair));
+void SpatialRPCService::IncrementAckedRPCID(Worker_EntityId EntityId, ERPCType Type)
+{
+	if (Type != ERPCType::NetMulticast)
```
[the remark this answers is not in the document]
Comment (MatthewSandfordImprobable) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialRPCService.cpp:483]: Pretty sure this can be called on multicast rpcs. I will return early though.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp (hunk 25 above)
[the remark this answers is not in the document]
Comment (MatthewSandfordImprobable) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:1974]: This is defo a warning. This could happen if we have authority over an actor upon receiving the RPC. The RPC can not be processed because we are waiting on an unresolved argument in a previous RPC. We then lose authority over the RPC actor. 

So yes we are dropping all RPC in the queue for non auth actors. 

"can we do the string interpolation" Sorry I am not sure what you mean here?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialRPCService.cpp [hunk 27]
```diff
@@ -464,13 +473,30 @@ void SpatialRPCService::ExtractRPCsForType(Worker_EntityId EntityId, ERPCType Ty
...
+	{
+		EntityRPCType EntityTypePair = EntityRPCType(EntityId, Type);
 
-			RPCRingBufferUtils::WriteAckToSchema(EndpointObject, Type, LastProcessedRPCId);
+		uint64* LastAckedRPCId = LastAckedRPCIds.Find(EntityTypePair);
+		if (LastAckedRPCId == nullptr)
+		{
+			UE_LOG(LogSpatialRPCService, Warning, TEXT("SpatialRPCService::IncrementAckedRPCID: Could not find last acked RPC id. Entity: %lld, RPC type: %s"), EntityId, *SpatialConstants::RPCTypeToString(Type));
```
[the remark this answers is not in the document]
Comment (MatthewSandfordImprobable) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialRPCService.cpp:490]: I am not sure. I thought I was seeing a case where this was happening but have not reproduce. I added this code more as a - "well if this does happen at least we have a log".

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp (hunk 25 above)
[the remark this answers is not in the document]
Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:1974]: In the case you just described it shouldn't be warning though as the user has done nothing wrong and everything will work. The next work will pick up the rpc correctly. This is just some internal book keeping at that point.
Might be easier to discuss offline.

[the remark this answers is not in the document]
Comment (MatthewSandfordImprobable) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:1974]: Ahh yes, you are right. There is no need to warn.
